webapi/Skills/ChatSkills/ChatSkill.cs

```diff
@@ -394,59 +446,42 @@ private async Task<ChatMessage> GetChatResponseAsync(string chatId, string userI
 
         // Query relevant semantic and document memories
         await this.UpdateBotResponseStatusOnClientAsync(chatId, "Extracting semantic and document memories", cancellationToken);
-        var chatMemoriesTokenLimit = (int)(remainingToken * this._promptOptions.MemoriesResponseContextWeight);
-        var documentContextTokenLimit = (int)(remainingToken * this._promptOptions.DocumentContextWeight);
+        var chatMemoriesTokenLimit = (int)(remainingTokenBudget * this._promptOptions.MemoriesResponseContextWeight);
+        var documentContextTokenLimit = (int)(remainingTokenBudget * this._promptOptions.DocumentContextWeight);
 
         string[] tasks = await Task.WhenAll<string>(
             this._semanticChatMemorySkill.QueryMemoriesAsync(userIntent, chatId, chatMemoriesTokenLimit, this._kernel.Memory, cancellationToken),
             this._documentMemorySkill.QueryDocumentsAsync(userIntent, chatId, documentContextTokenLimit, this._kernel.Memory, cancellationToken)
         );
 
+        // Add chat and document memories as additional context.
+        var chatContextTokenCount = 0;
         var chatMemories = tasks[0];
         var documentMemories = tasks[1];
-
-        // Fill in the chat history if there is any token budget left
         var chatContextComponents = new List<string>() { chatMemories, documentMemories };
-        var chatContextText = string.Join("\n\n", chatContextComponents.Where(c => !string.IsNullOrEmpty(c)));
-        var chatHistoryTokenLimit = remainingToken - TokenUtilities.TokenCount(chatContextText) - TokenUtilities.TokenCount(planResult);
+        foreach (var contextComponent in chatContextComponents.Where(c => !string.IsNullOrEmpty(c)))
+        {
+            promptTemplate.AddSystemMessage(contextComponent);
+            chatContextTokenCount += TokenUtilities.GetContextMessageTokenCount(AuthorRole.System, contextComponent);
+        }
 
+        // Fill in the chat history with remaining token budget.
         string chatHistory = string.Empty;
+        var chatHistoryTokenBudget = remainingTokenBudget - chatContextTokenCount - TokenUtilities.GetContextMessageTokenCount(AuthorRole.System, planResult);
 
-        // Append the chat history, if allowed.
-        if (chatHistoryTokenLimit > 0)
-        {
-            await this.UpdateBotResponseStatusOnClientAsync(chatId, "Extracting chat history", cancellationToken);
-            chatHistory = await this.ExtractChatHistoryAsync(chatId, chatHistoryTokenLimit, cancellationToken);
-            chatContext.ThrowIfFailed();
-            chatContextText = $"{chatContextText}\n{chatHistory}";
-        }
+        // Append previous messages
+        await this.UpdateBotResponseStatusOnClientAsync(chatId, "Extracting chat history", cancellationToken);
+        chatHistory = await this.GetAllowedChatHistoryAsync(chatId, chatHistoryTokenBudget, promptTemplate, cancellationToken);
+        chatContext.ThrowIfFailed();
```

Comment:
Do we need this?

Reply:
Need what?

Reply:
`chatContext.ThrowIfFailed();` None of the operations above touch the context.

Reply:
Oh you're right, removed